core/src/main/java/org/infinispan/functional/impl/EntryViews.java

```diff
@@ -32,81 +35,106 @@ private EntryViews() {
    }
 
    public static <K, V> ReadWriteEntryView<K, V> readWrite(CacheEntry<K, V> entry, V prevValue, Metadata prevMetadata) {
-      return new EntryAndPreviousReadWriteView<>(entry, prevValue, prevMetadata);
+      return new EntryAndPreviousReadWriteView<>(entry, prevValue, prevMetadata, new CacheEncoders());
+   }
+
+   public static <K, V> ReadWriteEntryView<K, V> readWrite(CacheEntry<K, V> entry, V prevValue, Metadata prevMetadata, CacheEncoders cacheEncoders) {
+      return new EntryAndPreviousReadWriteView<>(entry, prevValue, prevMetadata, cacheEncoders);
    }
 
    public static <K, V> ReadEntryView<K, V> noValue(K key) {
-      return new NoValueReadOnlyView<>(key);
+      return new NoValueReadOnlyView<>(key, new CacheEncoders());
+   }
+
+   public static <K, V> ReadEntryView<K, V> noValue(K key, CacheEncoders cacheEncoders) {
+      return new NoValueReadOnlyView<>(key, cacheEncoders);
    }
 
    /**
     * For convenience, a lambda might decide to return the entry view it
     * received as parameter, because that makes easy to return both value and
     * meta parameters back to the client.
-    *
+    * <p>
     * If the lambda function decides to return an view, launder it into an
     * immutable view to avoid the user trying apply any modifications to the
     * entry view from outside the lambda function.
-    *
+    * <p>
     * If the view is read-only, capture its data into a snapshot from the
     * cached entry and avoid changing underneath.
     */
    @SuppressWarnings("unchecked")
-   public static <R> R snapshot(R ret) {
+   public static <R> R snapshot(R ret, CacheEncoders cacheEncoders) {
       if (ret instanceof EntryBackedReadWriteView) {
          EntryBackedReadWriteView view = (EntryBackedReadWriteView) ret;
-         return (R) new ReadWriteSnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadWriteSnapshotView(view.key(), view.find().orElse(null), view.entry.getMetadata());
       } else if (ret instanceof EntryAndPreviousReadWriteView) {
          EntryAndPreviousReadWriteView view = (EntryAndPreviousReadWriteView) ret;
-         return (R) new ReadWriteSnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadWriteSnapshotView(view.key(), view.getCurrentValue(), view.entry.getMetadata());
       } else if (ret instanceof EntryBackedReadOnlyView) {
          EntryBackedReadOnlyView view = (EntryBackedReadOnlyView) ret;
-         return (R) new ReadOnlySnapshotView(view.key(), view.entry.getValue(), view.entry.getMetadata());
+         return (R) new ReadOnlySnapshotView(view.key(), view.find().orElse(null), view.entry.getMetadata());
+      } else if (ret instanceof NoValueReadOnlyView) {
+         NoValueReadOnlyView view = (NoValueReadOnlyView) ret;
+         return (R) new ReadOnlySnapshotView(view.key(), null, null);
       }
 
-      return ret;
+      return (R) cacheEncoders.valueFromStorage(ret);
```

Comment:
I remembered I did the conversion there too for some reason, some tests not passing, event if at first I didn't put anything there. But this is probably a wrong solution I did because the code was not working properly somewhere else. I'm going to remove it and see what it happens